### DISCUSSION TITLE
How to make a custom action space?

### DISCUSSION THREAD

**Discussion by KamatMayur** [google/brax#566] [General]
I want to use brax and make an environment where the action space isn't the control signal for the actuators but rather the joint configurations (qpos) of all the non root joints and then use this as a setpoint for PD control to generate the ctrl signal. I cannot find anyway to do this. The base env class does have a property called action_size but it has no setter. So cannot change it. It automatically gets set to the ctrl size from the xml model. This is very inconvenient and it would have been better if the user  had more control over the environment action space. Please let me know if there is a way to use a custom action space. Or please consider adding a feature to do so! 

Thank you

**Comment by btaba** [comment]
Hi @KamatMayur , can you just override action_size? You can also modify the XML to add more [actuators](https://mujoco.readthedocs.io/en/stable/XMLreference.html#actuator).

**Reply by KamatMayur** [reply]
> Hi @KamatMayur , can you just override action_size? You can also modify the XML to add more []([actuators](https://mujoco.readthedocs.io/en/stable/XMLreference.html#actuator).)

No that is not what I'm looking for. I am aware of adding more actuators but what I'm tryna say is that the action space is fixed to only output the ctrl signals for the actuators. And also the action space is bounded. And the bounds are determined by the ctrl bounds (-1,1) I guess by default although can be changed in the xml model. But I am trying to change the action space from control signal to set points for a PD controller.  The set points could be qpos values. So the action space should be the qpos. So I'm thinking there is no way to do this in brax yet. I thought of creating a custom environment from scratch using mjx but I'm facing some weird issues that I've highlited in the following post.
[Issue](https://github.com/google-deepmind/mujoco/issues/2252)

**Comment by WhyxyxIsNotAnAvailableName** [comment]
Have you solved your problem please, I have recently come across the same answer, I would be very grateful if you could give me the relevant answer
